students/rob_sanchez/Lesson 6/mailroom_pt4.py

```diff
@@ -0,0 +1,169 @@
+#!/usr/bin/env python3
+import sys
+import re
+
+
+# List of donors and donation amounts
+donor_dict = {"Tom Cruise": [100, 200, 300],
+              "Michael Jordan": [1300],
+              "Katy Perry": [4500, 1500],
+              "Adam Sandler": [500, 2400]}
+
+
+def main():
+    # Menu options
+    options = {1: send_thank_you,
+               2: create_report,
+               3: send_letters,
+               4: sys.exit}
+    prompt = "\nChoose an action:\n"
+    menu_sel = ("\n1 - Send a Thank You\n2 - Create a Report\n"
+                "3 - Send letters to everyone\n4 - Quit\n")
+
+    # User selection
+    while True:
+        try:
+            user_selection = input(prompt + menu_sel)
+            options.get(int(user_selection))()
+        except ValueError:
+            print("\nPlease select a numeric value...")
+        except TypeError:
+            print("\nOption {} is invalid. Try again...".format(user_selection))
+
+
+# Sends a thank you email to the selected donor
+def send_thank_you():
+    # Holds a history of the donor's name and donation amount
+    history = {}
+
+    # Get name of donor
+    donor_name = get_donor_name()
+
+    # Display list of donors when user types "list"
+    while donor_name.lower() == "list":
+        print ("\nList of Donors: " + get_formatted_values(list(donor_dict)))
+        donor_name = get_donor_name()
+
+    # Get donation amount
+    amt_input = get_donation_amount()
+
+    add_donor(donor_name, float(amt_input))
+
+    history["donor_name"] = donor_name
+    history["amount"] = amt_input
+
+    print(send_email(history))
+
+
+# Creates a summary report of the donations
+def create_report():
+    headers = ["Donor Name", "Total Given", "Num Gifts", "Average Gift"]
+    str_format = "{:<30} ${:>17} {:>16} ${:>14}"
+    print(("\n{:<30} | {:^15} | {:^15} | {:^15}").format(*headers))
+    print("-"*82)
+
+    donor_total = get_donor_totals(donor_dict)
+    num_gifts = get_num_gifts(donor_dict)
+    avg_gift = get_averages(donor_dict)
```

Comment:
I understand this was done for unit testing purposes.. but this can be quite expensive for a large data set (3 additional for loops).